README
ClutterGegl
============

A ClutterActor showing the rendered content of a GeglNode.

 - Multiple ClutterActors for a single GeglNode.
 - ClutterGegl vies on multiple nodes in a graph.
 - Automatic updates of views when properties/wiring of GEGL graph changes.

Please send feedback on ClutterGegl to pippin@gimp.org or pippin@o-hand.com.


ChangeLog
2008-08-05  Øyvind Kolås  <pippin@o-hand.com>

	Bumped version of clutter-gegl to 0.8.

	* configure.ac: incremented version.
	* Makefile.am: install in 0.8 directories.
	* README: reworded.
	* clutter-gegl.pc.in: fixed paths to reflect 0.7

2008-06-11  Øyvind Kolås  <pippin@gimp.org>

	* clutter-gegl.[ch]: copmuted the desired render size from the nodes
	bounding box for the default simple constructor.
	* examples/Makefile.am: made the makefile build all .c files into
	individual binaries.
	* examples/parsexml.c: (main): set the stage size to the actor size.

2008-06-11  Øyvind Kolås  <pippin@gimp.org>

	* Makefile.am: install in 0.7 directories.
	* configure.ac: update clutter requirement to 0.7
	* clutter-gegl.pc.in: fixed paths to reflect 0.7

2008-03-03  Øyvind Kolås  <pippin@o-hand.com>

	* Makefile.am: create clutter-gegl-x.x.pc from clutter-gegl.pc

	* configure.ac:
	* clutter-gegl.pc.in: updated clutter dependency to 0.6, updated
	clutter-gegl version to 0.6 as well.

2008-03-03  Øyvind Kolås  <pippin@o-hand.com>

	* examples/parsexml.c: updated default composition to match current
	GEGL default composition.
	* clutter-gegl.c: (computed_event): reindented code.

2008-01-10  Øyvind Kolås     <pippin@o-hand.com>

	* Imported into svn.

2007-10-20  Øyvind Kolås     <pippin@gimp.org>

	* Duplicated the clutter-cairo tree, and adapted it to render for
	  GEGL instead of cairo.

2007-08-15  Emmanuele Bassi  <ebassi@openedhand.com>

	* configure.ac:
	* Makefile.am:
	* clutter-gegl.pc.in: Bump up to 0.5.0 and require clutter-0.5
	as well.

2007-08-07  Emmanuele Bassi  <ebassi@openedhand.com>

	* configure.ac: Bump up to 0.4.0.

	* clutter-gegl.pc.in: Require clutter-0.4.

2007-08-01  Emmanuele Bassi  <ebassi@openedhand.com>

	* clutter-gegl.c: Now that we ensure that we are not accessing
	the gegl_surface_t object when the texture actor has been destroyed
	before the gegl_t context, we can destroy the surface when disposing
	of the texture.

2007-08-01  Emmanuele Bassi  <ebassi@openedhand.com>

	* clutter-gegl.c: Attach the ClutterGegl texture actor to the
	surface and use clutter_surface_finish() to dispose of the
	gegl_surface_t we use to generate the gegl context. This plugs
	a reference count bug in the bindings where the context is not
	destroyed before the end of the program.

2007-07-30  Emmanuele Bassi  <ebassi@openedhand.com>

	* clutter-gegl.[ch]: Add clutter_gegl_set_source_color(), a
	utility function for setting a ClutterColor as the source color
	for a Gegl context.

2007-07-17  Emmanuele Bassi  <ebassi@openedhand.com>

	* Makefile.am: Install the headers into the clutter-0.x include
	directory, so that each integration library header files are
	self-contained into the include directory of the clutter version
	they are linking to.

	* clutter-gegl.pc.in: Update template, and require gegl.

2007-07-06  Matthew Allum  <mallum@openedhand.com>

	* clutter-gegl.pc.in:
	* configure.ac:
	Fix pc and library versioning (also bump to 0.3.1)

2007-06-29  Matthew Allum  <mallum@openedhand.com>

	* configure.ac:
	Bump up version to 0.3

2007-06-11  Neil J. Patel  <njp@o-hand.com>

	* clutter-gegl.c: (gegl_destroy_func):
	* examples/clock.c: (main):
	* examples/flowers.c: (main):
	Update to work with clutter-0.3 

	* configure.ac:
	Update to check for clutter-0.3
	Update for 0.2 release.

================== 2007-03-27 0.1 Release ===============================

2007-03-27  Matthew Allum  <mallum@openedhand.com>

	* clutter-gegl.c: (gegl_destroy_func):
	Remove old commented out code
	* clutter-gegl.doap:
	Update for 0.1 Release

2007-03-25  Emmanuele Bassi  <ebassi@openedhand.com>

	* examples/flowers.c: Use less flowers.

	* examples/clock.c: Remove C99-isms; get the stage width and
	height from the actor instead of hardcoding it; quit on button
	press.

	* clutter-gegl.c: Remove useless chain up in the class
	initialisation code: we are deriving from ClutterTexture already;
	add G_PARAM_STATIC_NICK to the parameters flags; check for
	endianess when de-multiplying the alpha of the surface before
	feeding it to GL.

2007-03-22  Matthew Allum  <mallum@openedhand.com>

	* clutter-gegl.pc.in:
	Bump up to require clutter-0.2

2007-03-22  Matthew Allum  <mallum@openedhand.com>

	* clutter-gegl.c: 
	* clutter-gegl.h:
	Simplify API to just two calls via hooking into gegl_destroy()
	as to then sync the image surface to GL texture.
	* configure.ac:
	Require gegl 1.4 (For gegl_set_user_data ())
	* examples/clock.c: 
	* examples/flowers.c:
	Update for newer API.

2006-11-30  Matthew Allum  <mallum@openedhand.com>

	* clutter-gegl.doap:
	Add a doap file

2006-09-15  Matthew Allum  <mallum@openedhand.com>

	* Makefile.am:
	* examples/Makefile.am:
	* examples/flowers.c:
	Misc fixes from Bastien Nocera, see #153.

2006-08-27  Matthew Allum  <mallum@openedhand.com>

	* clutter-gegl.pc.in:
	Reference clutter-1.0 in .pc rather than just clutter

2006-08-26  Matthew Allum  <mallum@openedhand.com>

	* configure.ac:
	* examples/Makefile.am:
	* Makefile.am:
	* clock.c:
	* flowers.c:
	Add examples to autofoo.

2006-08-26  Matthew Allum  <mallum@openedhand.com>

	* Makefile:
	* Makefile.am:
	* autogen.sh:
	* clutter-gegl.pc.in:
	* configure.ac:
	Add initial autofoo

2006-08-15  Matthew Allum  <mallum@openedhand.com>

	* clutter-gegl.c: (clutter_gegl_end):
	Add another de premull method ( no difference )
	* flowers.c:
	Add some alpha.. Fix color randomness selection.

2006-08-14  Matthew Allum  <mallum@openedhand.com>

	* flowers.c: (make_flower_actor):
	lower tolerance value for better quality.

2006-08-14  Matthew Allum  <mallum@openedhand.com>

	* clutter-gegl.c: (clutter_gegl_new), (clutter_gegl_end):
	Convert Gegl surface data correctly to pixbuf and un pre-mul
        alpha.
	* flowers.c: (make_flower_actor), (tick):
	Minor tweaks to flower generation.

2006-08-13  Matthew Allum  <mallum@openedhand.com>

	* Makefile:
	* flowers.c:
	Add a new gegl demo/toy drawing 10's of randomly generated 

clutter-gegl.c
/*
 * ClutterGegl.
 *
 * An simple Clutter GEGL viewer actor:
 *
 * Created By Øyvind Kolås <pippin@gimp.org>/<pippin@o-hand.com> (c) 2007,
 * based on on Clutter cairo, authored By Matthew Allum
 * <mallum@openedhand.com> (C) 2006.
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 */

/**
 * SECTION:clutter-gegl
 * @short_description: Actor for displaying
 *
 * #ClutterGegl is a #ClutterTexture that displays text.
 */
#include "clutter-gegl.h"
#include <string.h>
#include <limits.h>
#include <stdlib.h>
#include <math.h>

G_DEFINE_TYPE (ClutterGegl, clutter_gegl, CLUTTER_TYPE_TEXTURE);

enum
{
  PROP_0,
  PROP_RENDER_WIDTH,
  PROP_RENDER_HEIGHT,
  PROP_NODE,
  PROP_X,
  PROP_Y,
  PROP_SCALE
};

#define CLUTTER_GEGL_GET_PRIVATE(obj) \
  (G_TYPE_INSTANCE_GET_PRIVATE ((obj), CLUTTER_TYPE_GEGL, ClutterGeglPrivate))

struct _ClutterGeglPrivate
{
  GeglNode      *node;
  guchar        *buffer;
  gint           width, height;
  gint           x, y;
  
  gdouble        scale;

  guint          monitor_id;
  GeglProcessor *processor;
};

void
gegl_view_repaint (ClutterGegl *view);
static void clutter_gegl_render (ClutterGegl *view);

static void
computed_event (GeglNode      *self,
                GeglRectangle *rect,
                ClutterGegl   *view)
{
  ClutterGeglPrivate  *priv;
  guchar              *buffer;

  g_assert (CLUTTER_IS_ACTOR (view));

  priv = CLUTTER_GEGL_GET_PRIVATE (view);

  if (!priv->node)
    return;

  buffer = g_malloc (rect->width * rect->height * 4);
  gegl_node_blit (priv->node, priv->scale,
                  rect,
                  babl_format ("R'G'B'A u8"),
                  buffer,
                  rect->width * 4,
                  GEGL_BLIT_CACHE);

  clutter_texture_set_area_from_rgb_data (CLUTTER_TEXTURE (view),
                                          buffer,
                                          TRUE,
                                          rect->x, rect->y,
                                          rect->width, rect->height,
                                          rect->width * 4,  /* rowstride */
                                          4,                /* bpp   */
                                          0,                /* flags */
                                          NULL);            /* error */
  g_free (buffer);
}

static void
invalidated_event (GeglNode      *self,
                   GeglRectangle *rect,
                   ClutterGegl   *view)
{
  gegl_view_repaint (view);
  clutter_actor_queue_redraw (CLUTTER_ACTOR (view));
}

static void
clutter_gegl_set_property (GObject      *object,
                           guint         prop_id,
                           const GValue *value,
                           GParamSpec   *pspec)
{
  ClutterGegl        *gegl;
  ClutterGeglPrivate *priv;

  gegl = CLUTTER_GEGL (object);
  priv = CLUTTER_GEGL_GET_PRIVATE (gegl);

  switch (prop_id)
    {
      case PROP_RENDER_WIDTH:
        priv->width = g_value_get_int (value);
        if (priv->buffer)
          {
            g_free (priv->buffer);
            priv->buffer = NULL;
          }
        clutter_actor_queue_redraw (CLUTTER_ACTOR(gegl));
        break;

      case PROP_RENDER_HEIGHT:
        priv->height = g_value_get_int (value);
        if (priv->buffer)
          {
            g_free (priv->buffer);
            priv->buffer = NULL;
          }
        clutter_actor_queue_redraw (CLUTTER_ACTOR(gegl));
        break;

    case PROP_NODE:
      if (priv->node)
        {
          g_object_unref (priv->node);
        }

      if (g_value_get_object (value))
        {
          priv->node = GEGL_NODE (g_value_dup_object (value));

          g_signal_connect_object (priv->node, "computed",
                                   G_CALLBACK (computed_event),
                                   gegl, 0);
          g_signal_connect_object (priv->node, "invalidated",
                                   G_CALLBACK (invalidated_event),
                                   gegl, 0);
          clutter_actor_queue_redraw (CLUTTER_ACTOR(gegl));
        }
      else
        {
          priv->node = NULL;
        }
      break;
    case PROP_X:
      priv->x = g_value_get_int (value);
      clutter_actor_queue_redraw (CLUTTER_ACTOR (gegl));
      break;
    case PROP_Y:
      priv->y = g_value_get_int (value);
      clutter_actor_queue_redraw (CLUTTER_ACTOR (gegl));
      break;
    case PROP_SCALE:
      priv->scale = g_value_get_double (value);
      clutter_actor_queue_redraw (CLUTTER_ACTOR (gegl));
      break;

      default:
        G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
        break;
    }
}

static void
clutter_gegl_get_property (GObject    *object,
                           guint       prop_id,
                           GValue     *value,
                           GParamSpec *pspec)
{
  ClutterGegl        *gegl;
  ClutterGeglPrivate *priv;

  gegl = CLUTTER_GEGL (object);
  priv = CLUTTER_GEGL_GET_PRIVATE (gegl);

  switch (prop_id)
    {
      case PROP_RENDER_WIDTH:
        g_value_set_int (value, priv->width);
        break;

      case PROP_RENDER_HEIGHT:
        g_value_set_int (value, priv->height);
        break;

      case PROP_NODE:
        g_value_set_object (value, priv->node);
        break;

      case PROP_X:
        g_value_set_int (value, priv->x);
        break;

      case PROP_Y:
        g_value_set_int (value, priv->y);
        break;

      case PROP_SCALE:
        g_value_set_double (value, priv->scale);
        break;

      default:
        G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
        break;
    }
}

static void
clutter_gegl_dispose (GObject *object)
{
  ClutterGegl         *self = CLUTTER_GEGL (object);
  ClutterGeglPrivate  *priv;

  priv = CLUTTER_GEGL_GET_PRIVATE (self);

  if (priv->node)
    {
      g_object_unref (priv->node);
      priv->node = NULL;
    }
  if (priv->processor)
    {
      gegl_processor_destroy (priv->processor);
      priv->processor = NULL;
    }
  if (priv->monitor_id)
    {
      g_source_remove (priv->monitor_id);
      priv->monitor_id = 0;
    }

  G_OBJECT_CLASS (clutter_gegl_parent_class)->dispose (object);
}

static void
clutter_gegl_finalize (GObject *object)
{
  ClutterGegl         *self = CLUTTER_GEGL (object);
  ClutterGeglPrivate  *priv;

  priv = CLUTTER_GEGL_GET_PRIVATE (self);

  if (priv->buffer)
    {
      g_free (priv->buffer);
      priv->buffer = NULL;
    }
  G_OBJECT_CLASS (clutter_gegl_parent_class)->finalize (object);
}


static GObject*
clutter_gegl_constructor (GType                  type,
                          guint                  n_construct_properties,
                          GObjectConstructParam *construct_properties)
{
  ClutterGegl         *view;
  ClutterGeglPrivate  *priv;
  GObject             *obj;

  obj = G_OBJECT_CLASS (clutter_gegl_parent_class)->constructor
        (type, n_construct_properties, construct_properties);

  /* Now all of the object properties are set */

  view = CLUTTER_GEGL (obj);
  priv = CLUTTER_GEGL_GET_PRIVATE (view);

  /*g_return_val_if_fail (priv->width && priv->height, NULL);*/

  if (priv->width == 0 ||
      priv->height == 0)
    {
      GeglRectangle r = gegl_node_get_bounding_box (priv->node);
      priv->width = r.width;      
      priv->height = r.height;      
    }


  clutter_actor_set_size (CLUTTER_ACTOR (view), priv->width, priv->height);
  gegl_view_repaint (view);
  clutter_gegl_render (view);
  return obj;
}

static void
clutter_gegl_class_init (ClutterGeglClass *klass)
{
  GObjectClass *gobject_class = G_OBJECT_CLASS (klass);

  gobject_class->finalize     = clutter_gegl_finalize;
  gobject_class->dispose      = clutter_gegl_dispose;
  gobject_class->set_property = clutter_gegl_set_property;
  gobject_class->get_property = clutter_gegl_get_property;
  gobject_class->constructor  = clutter_gegl_constructor;

  g_type_class_add_private (gobject_class, sizeof (ClutterGeglPrivate));

#define PARAM_FLAGS    (G_PARAM_CONSTRUCT_ONLY | \
                        G_PARAM_STATIC_NAME | G_PARAM_STATIC_NICK | \
                        G_PARAM_STATIC_BLURB | \
                        G_PARAM_READABLE | G_PARAM_WRITABLE)

  g_object_class_install_property (gobject_class,
                                   PROP_RENDER_WIDTH,
                                   g_param_spec_int ("render-width",
                                                     "Render-Width",
                                                     "Base width of rectangle to render",
                                                     0, G_MAXINT,
                                                     0,
                                                     PARAM_FLAGS));
  g_object_class_install_property (gobject_class,
                                   PROP_RENDER_HEIGHT,
                                   g_param_spec_int ("render-height",
                                                     "Render-Height",
                                                     "Base height of rectangle to render",
                                                     0, G_MAXINT,
                                                     0,
                                                     PARAM_FLAGS));

  g_object_class_install_property (gobject_class, PROP_X,
                                   g_param_spec_int ("x0",
                                                     "X",
                                                     "X origin",
                                                     G_MININT, G_MAXINT, 0,
                                                     G_PARAM_CONSTRUCT |
                                                     G_PARAM_READWRITE));
  g_object_class_install_property (gobject_class, PROP_Y,
                                   g_param_spec_int ("y0",
                                                     "Y",
                                                     "Y origin",
                                                     G_MININT, G_MAXINT, 0,
                                                     G_PARAM_CONSTRUCT |
                                                     G_PARAM_READWRITE));
  g_object_class_install_property (gobject_class, PROP_SCALE,
                                   g_param_spec_double ("scale",
                                                        "Scale",
                                                        "Zoom factor",
                                                        0.0, 100.0, 1.00,
                                                        G_PARAM_CONSTRUCT |
                                                        G_PARAM_READWRITE));
  g_object_class_install_property (gobject_class, PROP_NODE,
                                   g_param_spec_object ("node",
                                                        "Node",
                                                        "The node to render",
                                                        G_TYPE_OBJECT,
                                                        G_PARAM_CONSTRUCT |
                                                        G_PARAM_READWRITE));



#undef PARAM_FLAGS
}

static void
clutter_gegl_init (ClutterGegl *self)
{
  ClutterGeglPrivate *priv;
  
  priv = CLUTTER_GEGL_GET_PRIVATE (self);
}

/**
 * clutter_gegl_new
 * @width: clutter gegl surface width
 * @height: clutter gegl surface height
 *
 * Creates a new #ClutterGegl texture.
 *
 * Return value: a #ClutterGegl texture
 */
ClutterActor*
clutter_gegl_new (guint width,
                  guint height)
{
  return g_object_new (CLUTTER_TYPE_GEGL,
                       "render-width", width,
                       "render-height", height,
                       NULL);
}


ClutterActor * clutter_gegl_new_from_gegl_node (GeglNode *node)
{
  GeglRectangle extent = gegl_node_get_bounding_box (node);
  return clutter_gegl_new_from_gegl_node_with_size (node, extent.width, extent.height);
}



ClutterActor * clutter_gegl_new_from_gegl_node_with_size (GeglNode *node,
                                                          guint     width,
                                                          guint     height)
{
  return g_object_new (CLUTTER_TYPE_GEGL,
                       "render-width", width,
                       "render-height", height,
                       "node", node,
                       NULL);
}

static void
clutter_gegl_render (ClutterGegl *view)
{
  /* FIXME: should be optimized and handle ROIs (and still keep the buffer around) */
  ClutterGeglPrivate  *priv;
  GeglRectangle        roi;
  guchar              *buffer;

  g_assert (CLUTTER_IS_ACTOR (view));
  priv = CLUTTER_GEGL_GET_PRIVATE (view);

  if (!priv->node)
    return;

  buffer = g_malloc (priv->width * priv->height * 4);

  roi.x = priv->x;
  roi.y = priv->y;
  roi.width = priv->width;
  roi.height = priv->height;

  gegl_node_blit (priv->node, priv->scale,
                  &roi,
                  babl_format ("R'G'B'A u8"),
                  buffer,
                  GEGL_AUTO_ROWSTRIDE,
                  GEGL_BLIT_CACHE|GEGL_BLIT_DIRTY);

  clutter_texture_set_from_rgb_data (CLUTTER_TEXTURE (view),
                                     buffer,
                                     TRUE,               /* has_alpha */
                                     priv->width,
                                     priv->height,
                                     priv->width * 4,   /*(4 * priv->width + 3) &~ 3,*/
                                     4,
                                     0,
                                     NULL);
  g_free (buffer);

  if (CLUTTER_ACTOR_IS_VISIBLE (CLUTTER_ACTOR (view)))
    clutter_actor_queue_redraw (CLUTTER_ACTOR (view));
}

/**
 * clutter_gegl_create
 * @gegl:  A #ClutterGegl texture.
 *
 * Creates a new gegl context #ClutterGegl texture.
 *
 * Return value: a newly created gegl context. Free with gegl_destroy()
 * when you are done drawing.
 */
ClutterActor * clutter_gegl_parse_xml (const gchar *xml)
{
  return NULL;
}

static gboolean
task_monitor (ClutterGegl *view)
{
  ClutterGeglPrivate *priv = CLUTTER_GEGL_GET_PRIVATE (view);
  if (priv->processor == NULL ||
      priv->monitor_id == 0)
    {
      priv->monitor_id = 0;
      return FALSE;
    }
  if (gegl_processor_work (priv->processor, NULL))
    return TRUE;

  priv->monitor_id = 0;
  return FALSE;
}

void
gegl_view_repaint (ClutterGegl *view)
{
  ClutterGeglPrivate *priv = CLUTTER_GEGL_GET_PRIVATE (view);
#if 0
  ClutterActor  *actor = CLUTTER_ACTOR (view);
  GeglRectangle  roi    = { priv->x / priv->scale, priv->y / priv->scale,
                            ceil(clutter_actor_get_width (actor) / priv->scale+1),
                            ceil(clutter_actor_get_height (actor) / priv->scale+1) };
#endif
  GeglRectangle  roi    = { 0, 0, priv->width, priv->height };

  if (priv->monitor_id == 0)
    {
      priv->monitor_id = g_idle_add_full (G_PRIORITY_LOW,
                                          (GSourceFunc) task_monitor, view,
                                          NULL);

      if (priv->processor == NULL)
        {
          if (priv->node)
            priv->processor = gegl_node_new_processor (priv->node, &roi);
        }
    }
  if (priv->processor)
    gegl_processor_set_rectangle (priv->processor, &roi);
}

GeglProcessor *gegl_view_get_processor (ClutterGegl *view)
{
  ClutterGeglPrivate *priv = CLUTTER_GEGL_GET_PRIVATE (view);
  return priv->processor;
}


clutter-gegl.h
/*
 * ClutterGegl.
 *
 * An simple Clutter GEGL viewer actor:
 *
 * Created By Øyvind Kolås <pippin@gimp.org>/<pippin@o-hand.com> (c) 2007,
 * based on on Clutter cairo, authored By Matthew Allum
 * <mallum@openedhand.com> (C) 2006.
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 */

#ifndef _HAVE_CLUTTER_GEGL_H
#define _HAVE_CLUTTER_GEGL_H

#include <glib-object.h>

#include <gdk-pixbuf/gdk-pixbuf.h>
#include <clutter/clutter-texture.h>

#include <gegl.h>

G_BEGIN_DECLS

#define CLUTTER_TYPE_GEGL    clutter_gegl_get_type ()

#define CLUTTER_GEGL(obj) \
  (G_TYPE_CHECK_INSTANCE_CAST ((obj), \
                               CLUTTER_TYPE_GEGL, ClutterGegl))

#define CLUTTER_GEGL_CLASS(klass) \
  (G_TYPE_CHECK_CLASS_CAST ((klass), \
                            CLUTTER_TYPE_GEGL, ClutterGeglClass))

#define CLUTTER_IS_GEGL(obj) \
  (G_TYPE_CHECK_INSTANCE_TYPE ((obj), \
                               CLUTTER_TYPE_GEGL))

#define CLUTTER_IS_GEGL_CLASS(klass) \
  (G_TYPE_CHECK_CLASS_TYPE ((klass), \
                            CLUTTER_TYPE_GEGL))

#define CLUTTER_GEGL_GET_CLASS(obj) \
  (G_TYPE_INSTANCE_GET_CLASS ((obj), \
                              CLUTTER_TYPE_GEGL, ClutterGeglClass))

typedef struct _ClutterGegl        ClutterGegl;
typedef struct _ClutterGeglClass   ClutterGeglClass;
typedef struct _ClutterGeglPrivate ClutterGeglPrivate;

struct _ClutterGegl
{
  ClutterTexture        parent;
#if 0
  /*< private >*/
  ClutterGeglPrivate   *priv;
#endif
};

struct _ClutterGeglClass
{
  /*< private >*/
  ClutterTextureClass parent_class;

  void (*_clutter_gegl_1)(void);
  void (*_clutter_gegl_2)(void);
  void (*_clutter_gegl_3)(void);
  void (*_clutter_gegl_4)(void);
};

GType
clutter_gegl_get_type                          (void) G_GNUC_CONST;

ClutterActor * clutter_gegl_parse_xml          (const gchar *xml);

ClutterActor * clutter_gegl_new_from_gegl_node (GeglNode    *node);

ClutterActor * clutter_gegl_new_from_gegl_node_with_size (GeglNode    *node,
                                                          guint        width,
                                                          guint        height);
                                                

G_END_DECLS

#endif


examples/parsexml.c
/*
 * ClutterGegl.
 *
 * An simple Clutter GEGL viewer actor:
 *
 * Created By Øyvind Kolås <pippin@gimp.org|o-hand.com> (c) 2007,
 * based on on Clutter cairo, authored By Matthew Allum
 * <mallum@openedhand.com> (C) 2006.
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 */

#include <clutter/clutter.h>
#include "clutter-gegl.h"
#include "math.h"

#define DEFAULT_COMPOSITION "<?xml version='1.0' encoding='UTF-8'?> <gegl> <node operation='crop'> <params> <param name='x'>0.000000</param> <param name='y'>0.000000</param> <param name='width'>400.000000</param> <param name='height'>290.000000</param> </params> </node> <node operation='over'> <node operation='shift'> <params> <param name='x'>50.000000</param> <param name='y'>113.000000</param> </params> </node> <node operation='dropshadow'> <params> <param name='opacity'>2.000000</param> <param name='x'>2.000000</param> <param name='y'>2.000000</param> <param name='radius'>3.000000</param> </params> </node> <node operation='text'> <params> <param name='string'>A graph based image processing and\ncompositing engine.</param> <param name='font'>Sans</param> <param name='size'>10.000000</param> <param name='color'>rgb(1.0000, 1.0000, 1.0000)</param> <param name='wrap'>-1</param> <param name='alignment'>0</param> <param name='width'>178</param> <param name='height'>26</param> </params> </node> </node> <node operation='over'> <node operation='shift'> <params> <param name='x'>132.000000</param> <param name='y'>180.000000</param> </params> </node> <node operation='dropshadow'> <params> <param name='opacity'>2.000000</param> <param name='x'>2.000000</param> <param name='y'>2.000000</param> <param name='radius'>3.000000</param> </params> </node> <node operation='text'> <params> <param name='string'>2000-2007 © Calvin Williamson, Caroline Dahloff, Manish Singh, Jay Cox, Daniel Rogers, Sven Neumann, Michael Natterer, Øyvind Kolås, Philip Lafleur, Dominik Ernst, Richard Kralovic, Kevin Cozens, Victor Bogado, Martin Nordholts, Geert Jordaens, Michael Schumacher, Garry R. Osgood and Jakub Steiner</param> <param name='font'>Sans</param> <param name='size'>10.458599</param> <param name='color'>rgb(1.0000, 1.0000, 1.0000)</param> <param name='wrap'>240</param> <param name='alignment'>0</param> <param name='width'>237</param> <param name='height'>130</param> </params> </node> </node> <node operation='over'> <node name='dropshadow' operation='dropshadow'> <params> <param name='opacity'>-4.487179</param> <param name='x'>3.750000</param> <param name='y'>2.500000</param> <param name='radius'>10.597015</param> </params> </node> <node operation='shift'> <params> <param name='x'>42.000000</param> <param name='y'>23.000000</param> </params> </node> <node operation='text'> <params> <param name='string'>GEGL</param> <param name='font'>Sans</param> <param name='size'>76.038217</param> <param name='color'>rgb(0.3093, 0.3910, 0.1680)</param> <param name='wrap'>-1</param> <param name='alignment'>0</param> <param name='width'>208</param> <param name='height'>89</param> </params> </node> </node> <node operation='whitebalance'> <params> <param name='high-a-delta'>0.365079</param> <param name='high-b-delta'>0.026511</param> <param name='low-a-delta'>0.000000</param> <param name='low-b-delta'>0.000000</param> <param name='saturation'>0.319548</param> </params> </node> <node operation='brightness-contrast'> <params> <param name='contrast'>0.229008</param> <param name='brightness'>0.000000</param> </params> </node> <node operation='fractal-explorer'> <params> <param name='width'>400</param> <param name='height'>290</param> <param name='fractaltype'>1</param> <param name='xmin'>-1.312500</param> <param name='xmax'>0.000000</param> <param name='ymin'>-0.609375</param> <param name='ymax'>0.562500</param> <param name='iter'>100</param> <param name='cx'>-0.820312</param> <param name='cy'>0.195312</param> <param name='redstretch'>1.000000</param> <param name='greenstretch'>1.000000</param> <param name='bluestretch'>1.000000</param> <param name='redmode'>1</param> <param name='greenmode'>1</param> <param name='bluemode'>0</param> <param name='redinvert'>false</param> <param name='greeninvert'>false</param> <param name='blueinvert'>false</param> <param name='ncolors'>256</param> <param name='useloglog'>false</param> </params> </node> </gegl>"

int
main (int argc, char **argv)
{
  GeglNode        *gegl;
  ClutterActor    *geglview, *stage;
  ClutterColor     stage_color = { 0x99, 0xcc, 0xff, 0xff };

  clutter_init (&argc, &argv);
  gegl_init (&argc, &argv);

  stage = clutter_stage_get_default ();
  g_signal_connect (stage, "button-press-event",
                    G_CALLBACK (clutter_main_quit),
                    NULL);

  clutter_stage_set_color (CLUTTER_STAGE (stage),
		           &stage_color);

  if (argv[1])
    {
      gegl = gegl_node_new_from_file (argv[1]);
    }
  else
    {
      gegl = gegl_node_new_from_xml (DEFAULT_COMPOSITION, "");
    }

  geglview = clutter_gegl_new_from_gegl_node (gegl);
  clutter_group_add (CLUTTER_GROUP(stage), geglview);
  clutter_actor_set_size (stage, clutter_actor_get_width (geglview),
                                 clutter_actor_get_height (geglview));

  clutter_actor_show_all (CLUTTER_ACTOR(stage));

  clutter_main();

  return 1;
}
